query-planner-wasm/src/lib.rs

```diff
@@ -4,31 +4,61 @@ use apollo_query_planner::{QueryPlanner, QueryPlanningOptions};
 use js_sys::JsString;
 use wasm_bindgen::prelude::*;
 
-static mut SCHEMA: Vec<String> = vec![];
-static mut DATA: Vec<QueryPlanner> = vec![];
+/// These vectors serve as repositories for the Schema string and QueryPlanner
+/// so that they don't get garbage collected. The JavaScript side gets the index
+/// into the vector, which it can use in `getQueryPlan`.
+static mut SCHEMA: Vec<Option<String>> = vec![];
+static mut DATA: Vec<Option<QueryPlanner>> = vec![];
 
+/// getQueryPlanner creates a QueryPlanner if needed, and returns an "id"
+/// for later use with `getQueryPlan`. Calling this multiple times with
+/// the same schema will only result in the schema being parsed once, and the
+/// QueryPlanner is reused.
 #[wasm_bindgen(js_name = getQueryPlanner)]
 pub fn get_query_planner(schema: JsString) -> usize {
+    let schema = String::from(schema);
     unsafe {
-        if SCHEMA.is_empty() {
-            SCHEMA.push(String::from(schema));
-            DATA.push(QueryPlanner::new(&SCHEMA[0]));
-        } else {
-            SCHEMA[0] = String::from(schema);
-            DATA[0] = QueryPlanner::new(&SCHEMA[0]);
+        for i in 0..SCHEMA.len() {
+            match &SCHEMA[i] {
+                Some(x) if x == &schema => {
+                    return i;
+                }
+                _ => (),
+            }
+        }
+        let id = SCHEMA.len();
+        SCHEMA.push(Some(schema));
+        DATA.push(Some(QueryPlanner::new(&SCHEMA[id].as_ref().unwrap())));
+        return id;
+    }
+}
+
+/// Drop a query planner (and associated Schema string) to free up memory.
+/// Most applications will have a single query planner that they use
+/// for the duration of the app's lifetime, but if you are working
```

Comment:
This sentence is a bit misleading; One of the ways in which Apollo Gateway can run is by polling for downstream schema changes, composing a new federated schema (csdl) and updating the runtime schema on the fly.

That fact brings to question if this is the right solution, because some people do in fact create a lot of new schemas during the runtime of the server by design.
I like that you've added the function to drop a planner, if we want to go with this approach, we'd have to make sure and drop the old planner whenever a schema updates (i.e. identify schema changes vs. additions).

At the root of the problem though, is that there is one global wasm module loaded, but there can be multiple ApolloGateway instances. We might want to solve this at the root of the problem, if it's possible, rather than adapting the global model to a local one.

Let me stew on this for a bit.

Generally though, the use of an index is indeed better than a pointer!